dozer-sql/expression/src/scalar/string.rs

```diff
@@ -272,6 +272,200 @@ pub(crate) fn evaluate_to_char(
+        Field::U128(u) => Ok(Field::String((((u % 256) as u8) as char).to_string())),
+        Field::Int(i) => Ok(Field::String((((i % 256) as u8) as char).to_string())),
+        Field::I128(i) => Ok(Field::String((((i % 256) as u8) as char).to_string())),
+        Field::Float(_)
+        | Field::Decimal(_)
+        | Field::Boolean(_)
+        | Field::String(_)
+        | Field::Text(_)
+        | Field::Date(_)
+        | Field::Timestamp(_)
+        | Field::Binary(_)
+        | Field::Json(_)
+        | Field::Point(_)
+        | Field::Duration(_)
+        | Field::Null => Err(Error::InvalidFunctionArgument {
+            function_name: ScalarFunctionType::Chr.to_string(),
+            argument_index: 0,
+            argument: value,
+        }),
+    }
+}
+
+pub fn validate_substr(args: &[Expression], schema: &Schema) -> Result<ExpressionType, Error> {
+    validate_num_arguments(2..4, args.len(), ScalarFunctionType::Substr)?;
+
+    if args.len() == 2 {
+        validate_arg_type(
+            &args[0],
+            vec![FieldType::String, FieldType::Text],
+            schema,
+            ScalarFunctionType::Substr,
+            0,
+        )?;
+        validate_arg_type(
+            &args[1],
+            vec![
+                FieldType::UInt,
+                FieldType::U128,
+                FieldType::Int,
+                FieldType::I128,
+            ],
+            schema,
+            ScalarFunctionType::Substr,
+            1,
+        )?;
+    } else {
+        validate_arg_type(
+            &args[0],
+            vec![FieldType::String, FieldType::Text],
+            schema,
+            ScalarFunctionType::Substr,
+            0,
+        )?;
+        validate_arg_type(
+            &args[1],
+            vec![
+                FieldType::UInt,
+                FieldType::U128,
+                FieldType::Int,
+                FieldType::I128,
+            ],
+            schema,
+            ScalarFunctionType::Substr,
+            1,
+        )?;
+        validate_arg_type(
+            &args[2],
+            vec![
+                FieldType::UInt,
+                FieldType::U128,
+                FieldType::Int,
+                FieldType::I128,
+            ],
+            schema,
+            ScalarFunctionType::Substr,
+            2,
+        )?;
+    }
+
+    let ret_type = FieldType::String;
+
+    Ok(ExpressionType::new(
+        ret_type,
+        false,
+        dozer_types::types::SourceDefinition::Dynamic,
+        false,
+    ))
+}
+
+pub(crate) fn evaluate_substr(
+    schema: &Schema,
+    arg: &mut Expression,
+    position: &mut Expression,
+    length: &mut Option<Box<Expression>>,
+    record: &Record,
+) -> Result<Field, Error> {
+    let arg_field = arg.evaluate(record, schema)?;
+    let arg_value = arg_field.to_string();
+
+    let position_field = position.evaluate(record, schema)?;
+    let position_result = position_field.to_i128();
```

Comment:
A string of i64::MAX length is about 8 exabytes, I believe, so let's just use u64 here, which tends to be faster for computations. Below, it's converted to `usize` anyway.